### DISCUSSION TITLE
ESP8266: How to configure 8 Input Channels (Optocoupler)

### DISCUSSION THREAD

**Discussion by Noschvie** [arendst/Tasmota#19844] [Q&A] [answered]
Hello
have a board with 8 optocopler connected to a ESP8266 MCU.
Which pins are useable for this?
D1 up to D7 are working, but for the 8th input?

![grafik](https://github.com/arendst/Tasmota/assets/48217146/fc47f1d8-96f7-4f8a-b00d-c38e584f512e)

```
Backlog switchmode1 1; switchmode2 1; switchmode3 1; switchmode4 1; switchmode5 1; switchmode6 1; switchmode7 1; switchmode8 1
rule1
on switch1#state=0 do publish stat/%topic%/switch1 ON endon on switch1#state=1 do Publish stat/%topic%/switch1 OFF endon
on switch2#state=0 do publish stat/%topic%/switch2 ON endon on switch2#state=1 do Publish stat/%topic%/switch2 OFF endon
on switch3#state=0 do publish stat/%topic%/switch3 ON endon on switch3#state=1 do Publish stat/%topic%/switch3 OFF endon
on switch4#state=0 do publish stat/%topic%/switch4 ON endon on switch4#state=1 do Publish stat/%topic%/switch4 OFF endon
rule1 1
rule2
on switch5#state=0 do publish stat/%topic%/switch5 ON endon on switch5#state=1 do Publish stat/%topic%/switch5 OFF endon
on switch6#state=0 do publish stat/%topic%/switch6 ON endon on switch6#state=1 do Publish stat/%topic%/switch6 OFF endon
on switch7#state=0 do publish stat/%topic%/switch7 ON endon on switch7#state=1 do Publish stat/%topic%/switch7 OFF endon
on switch8#state=0 do publish stat/%topic%/switch8 ON endon on switch8#state=1 do Publish stat/%topic%/switch8 OFF endon
rule2 1
```

Maybe an ESP32 shall be used ?
Thansks!

**Comment by sfromis** [comment]
"Usable" depends on what you mean by that, and what other things may be connected, like if it is a dev board with an USB/serial adapter. If not, you have gpio 1,3. Beware of pin state at boot time, restrictions apply, but gpio 15/16 are not unusable, while you still need to be careful. Gpio 2 also needs care, both for boot reasons, and also due to it being a led on typical dev boards. And gpio0 also has "special effects".
![randomnerdtutorials com esp8266-pinout-reference-gpios](https://github.com/arendst/Tasmota/assets/47082390/c52464fd-a444-4b30-8cc8-56ee4ed38859)

Of course, ESP32 gives more pins, but 8 should be feasible on ESP8266, if you are careful with the defails.

**Comment by Noschvie** [comment]
Ok, will take an ESP32 D1 Mini NodeMCU. Which PINs shall be used to that ?
Thanks!

**Reply by sfromis** [reply, marked as answer]
For ESP32, like for ESP8266, start with the "green" pins in the table, to the extent they are broken out on your board. It is not a given that all pins are easily available on any board.
https://randomnerdtutorials.com/esp32-pinout-reference-gpios/

**Comment by Noschvie** [comment]
Now I have taken a Wemos ESP32C3mini V2.1.0 with the configuration below and the 8 input channels are working.
![image](https://github.com/arendst/Tasmota/assets/48217146/0655a4a2-44f7-468a-a131-4678c71931eb)
